bayes/Calibration.py

```diff
@@ -0,0 +1,143 @@
+        Model to construct prior, likelihood and the posterior with the supplied Observed data
+        :param observed_data:
+        :return:
+        #TODO: Infer likelihood noise/hyperprior (atleast the MAP)
+        """
+        # --prior
+        # --- https://github.com/stan-dev/stan/wiki/Prior-Choice-Recommendations
+        # TODO: Incorporate prior_dist user choice here
+        if self.prior_dist == "Normal":
+            self.para_prior = dist.Normal(self.prior_hyperparameters[0], self.prior_hyperparameters[1])
+        if self.prior_dist == "Uniform":
+            self.para_prior = dist.Uniform(self.prior_hyperparameters[0], self.prior_hyperparameters[1])
+        if self.prior_dist == "Gamma":
+            self.para_prior = dist.Gamma(self.prior_hyperparameters[0], self.prior_hyperparameters[1])
+        if self.prior_dist == "Beta":
+            self.para_prior = dist.Beta(self.prior_hyperparameters[0], self.prior_hyperparameters[1])
+        para = pyro.sample("theta", self.para_prior)
+        # --likelihood
+        mean = self.forward_solve(self.fw_input, para)
+        # TODO: Incorporate noise model dist choice here, default is normal
+        # TODO: More involved noise model, with correlation structure
+        if self.obs_noise_dist == "Normal":
+            self.likelihood = dist.Normal(mean, self.obs_noise_parameters ** 2)
+        else:
+            raise NotImplementedError
+        pyro.sample("lkl", self.likelihood, obs=observed_data)
+
+        return para
+
+    def run(self, n, kernel=None):
+        """
+        Method to find approximate posterior with MCMC/VI
+        :param n:
+        :param kernel:
+        :return: posterior: [N,]
+        """
+        # TODO: Incorporate VI
+        if kernel == "NUTS":
+            kernel = NUTS(self.model)
+        if kernel == "HMC":
+            print('Using HMC kernel')
+            kernel = HMC(model=self.model, step_size=1, num_steps=4)
+
+        mcmc = MCMC(kernel, num_samples=n)
+        mcmc.run(self.observed_data)
+        posterior = mcmc.get_samples()['theta'].numpy()
+        # print(posterior)
+        mcmc.summary()
+        return posterior
+
+    def predict(self, posterior, new_input):
+        """
+        Method to get posterior predictive distribution.
+        :param posterior: p(theta|D) samples
+        :param new_input: New input to the solver. [Dict type] ('known_parameters': , 'sensors': ,'time_steps': )
+        :return: tilda_X: New unobserved data samples
+        """
+        size = np.size(new_input['sensors']) * new_input['time_steps']
+        tilda_X: ndarray = np.ndarray((np.size(posterior), size))
+        for i in range(0, np.size(posterior)):
+            theta = posterior[i]
+            mean = self.forward_solve(new_input, theta)
+            _dist = dist.Normal(mean, self.obs_noise_parameters)
+            tilda_X[i, :] = pyro.sample("pos", _dist)
+        return tilda_X
+
+    def visualize_prior_posterior(self, posterior):
```

Comment:
I would suggest to move the visualization into a separate class/file with well defined inputs and outputs.

Reply:
Yes WIP